proof/refine/ARM/Syscall_R.thy

```diff
@@ -351,50 +351,49 @@ lemma threadSet_tcbDomain_update_sch_act_wf[wp]:
-  sorry (*
-                 apply (rule corres_split_deprecated[OF _ gts_isRunnable_corres])
-                   apply simp
-                   apply (rule corres_split_deprecated[OF corres_when[OF refl]])
-                      apply (rule rescheduleRequired_corres)
-                     apply clarsimp
-                     apply (rule corres_when[OF refl])
-                     apply (rule tcbSchedEnqueue_corres)
-                    apply (wp hoare_drop_imps hoare_vcg_conj_lift | clarsimp| assumption)+
-          apply (clarsimp simp: etcb_relation_def)
-         apply ((wp hoare_vcg_conj_lift hoare_vcg_disj_lift | clarsimp)+)[1]
-        apply clarsimp
-        apply (rule_tac Q="\<lambda>_. valid_objs' and valid_queues' and valid_queues and
-          (\<lambda>s. sch_act_wf (ksSchedulerAction s) s) and tcb_at' tptr"
-          in hoare_strengthen_post[rotated])
-         apply (auto simp: invs'_def valid_state'_def sch_act_wf_weak st_tcb_at'_def o_def)[1]
-        apply (wp threadSet_valid_objs' threadSet_valid_queues'_no_state
-          threadSet_valid_queues_no_state
-          threadSet_pred_tcb_no_state | simp)+
-      apply (rule_tac Q = "\<lambda>r s. invs' s \<and> (\<forall>p. tptr \<notin> set (ksReadyQueues s p)) \<and> sch_act_simple s
-        \<and>  tcb_at' tptr s" in hoare_strengthen_post[rotated])
-       apply (clarsimp simp:invs'_def valid_state'_def valid_pspace'_def sch_act_simple_def)
-       apply (clarsimp simp:valid_tcb'_def)
-       apply (drule(1) bspec)
-       apply (clarsimp simp:tcb_cte_cases_def)
-       apply fastforce
-      apply (wp hoare_vcg_all_lift Tcb_R.tcbSchedDequeue_not_in_queue)+
-   apply clarsimp
-   apply (frule tcb_at_is_etcb_at)
-    apply simp+
-   apply (auto elim: tcb_at_is_etcb_at valid_objs'_maxDomain valid_objs'_maxPriority pred_tcb'_weakenE
-               simp: valid_sched_def valid_sched_action_def)
-  done *)
+  apply (rule stronger_corres_guard_imp)
+    apply (rule corres_split[OF gct_corres])
+      apply (rule corres_split[OF tcbSchedDequeue_corres])
+        apply (rule corres_split[OF threadset_corresT])
+             apply (clarsimp simp: tcb_relation_def)
+            apply (clarsimp simp: tcb_cap_cases_def)
+           apply (clarsimp simp: tcb_cte_cases_def)
+          apply (rule corres_split[OF isSchedulable_corres])
+            apply (rule corres_split[OF corres_when[OF _ tcbSchedEnqueue_corres]], simp)
+              apply (rule corres_when[OF _ rescheduleRequired_corres], simp)
+             apply (wpsimp wp: hoare_drop_imp hoare_vcg_if_lift2 thread_set_weak_valid_sched_action
+                               threadSet_valid_tcbs' threadSet_vrq_inv threadSet_vrq'_inv
+                               threadSet_valid_queues_no_state threadSet_valid_queues'_no_state
+                         simp: valid_tcb_domain_update)+
+      apply (clarsimp cong: conj_cong)
+      apply (rule hoare_vcg_conj_lift, strengthen valid_tcb'_tcbDomain_update, wpsimp)
```

Comment:
It's difficult to judge without seeing the proof state, but I wonder if `valid_tcb'_tcbDomain_update` would work as a `simp:` rule here if you dropped the `forall` from the rule.

Reply:
Ignore that, I misread the rule. (After seeing `valid_tcb_domain_update`, I assumed `valid_tcb'_tcbDomain_update` was a conditional equality rather than an implication).